osu.Game/Screens/Select/Carousel/CarouselBeatmap.cs

```diff
@@ -55,12 +55,31 @@ public override void Filter(FilterCriteria criteria)
 
             match &= !criteria.UserStarDifficulty.HasFilter || criteria.UserStarDifficulty.IsInRange(BeatmapInfo.StarRating);
 
-            if (match && criteria.SearchTerms.Length > 0)
+            if (!match)
             {
-                string[] terms = BeatmapInfo.GetSearchableTerms();
+                Filtered.Value = !match;
+                return;
+            }
```

Comment:
If we're going in here and optimising, you could probably get a lot of free wins out of early-returning everywhere where `&=` is used above, but the amount of early-returns would get very ugly.

I'd say that extracting the computation of `match` to a bool-returning method and using early-returns of `false` on each criteria check is going to clean up much better than this sort of ugly early-return. I would risk the guess that at that point you could probably even get similar results to the elimination of the `SearchableTerms` iterator allocs without actually eliminating them, because most items wouldn't even *reach* trying to iterate over it.

Reply:
turns out early-returning everywhere doesn't give significant benefit. Probably because most of the time only one filter is active at the same time.

Reply:
I've done a bit of rearranging on this method, and I think more can be done to improve code quality, but I don't want to overthink it as the final solution involves a rewrite of most of this to work in realm-query-space.